pandas/core/window_/indexers.py

```diff
@@ -0,0 +1,181 @@
+import abc
+from typing import Optional, Sequence, Tuple, Union
+
+import numpy as np
+
+from pandas.tseries.offsets import DateOffset
+
+BeginEnd = Tuple[np.ndarray, np.ndarray]
+
+
+class BaseIndexer(abc.ABC):
+    """Base class for window bounds calculations"""
+
+    def __init__(
+        self,
+        index=None,
+        offset: Optional[Union[str, DateOffset]] = None,
+        keys: Optional[Sequence[np.ndarray]] = None,
+    ):
+        """
+        Parameters
+        ----------
+        index : , default None
+            pandas index to reference in the window bound calculation
+
+        offset: str or DateOffset, default None
+            Offset used to calcuate the window boundary
+
+        keys: np.ndarray, default None
+            Additional columns needed to calculate the window bounds
+
+        """
+        self.index = index
+        self.offset = offset
+        self.keys = keys
```

Comment:
are we expecting people to instantiate BaseIndexer, or should make that completely abstract and use CustomIndexer? (for a custom indexer)

Reply:
I was expecting users to just directly subclass `BaseIndexer` to create a "CustomIndexer". So yes, expecting users to directly instantiate it.

Reply:
right I don't think that's a good idea, BaseIndexer should be completely abstract and folks should subclass it and/or CustomIndexer (new kind of empty class that just inherits BaseIndexer). We have this kind of structure for EA's (now)